File: app/src/main/java/fr/free/nrw/commons/mwapi/UserClient.java
package fr.free.nrw.commons.mwapi;

import fr.free.nrw.commons.wikidata.mwapi.MwQueryResponse;
import fr.free.nrw.commons.wikidata.mwapi.MwQueryResult;
import fr.free.nrw.commons.wikidata.mwapi.UserInfo;
import fr.free.nrw.commons.utils.DateUtil;

import java.util.Collections;
import java.util.Date;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.Single;

public class UserClient {
    private final UserInterface userInterface;

    @Inject
    public UserClient(UserInterface userInterface) {
        this.userInterface = userInterface;
    }

    /**
     * Checks to see if a user is currently blocked from Commons
     *
     * @return whether or not the user is blocked from Commons
     */
    public Single<Boolean> isUserBlockedFromCommons() {
        return userInterface.getUserBlockInfo()
                .map(MwQueryResponse::query)
                .map(MwQueryResult::userInfo)
                .map(UserInfo::blockexpiry)
                .map(blockExpiry -> {
                    if (blockExpiry.isEmpty())
                        return false;
                    else if ("infinite".equals(blockExpiry))
                        return true;
                    else {
                        Date endDate = DateUtil.iso8601DateParse(blockExpiry);
                        Date current = new Date();
                        return endDate.after(current);
                    }
                }).single(false);
    }
}

File: app/src/main/java/fr/free/nrw/commons/mwapi/UserClient.kt
package fr.free.nrw.commons.mwapi

import fr.free.nrw.commons.utils.DateUtil
import fr.free.nrw.commons.wikidata.mwapi.MwQueryResponse
import fr.free.nrw.commons.wikidata.mwapi.MwQueryResult
import fr.free.nrw.commons.wikidata.mwapi.UserInfo
import io.reactivex.Single
import java.text.ParseException
import java.util.Date
import javax.inject.Inject

class UserClient @Inject constructor(private val userInterface: UserInterface) {
    /**
     * Checks to see if a user is currently blocked from Commons
     *
     * @return whether or not the user is blocked from Commons
     */
    fun isUserBlockedFromCommons(): Single<Boolean> =
        userInterface.getUserBlockInfo()
            .map(::processBlockExpiry)
            .single(false)

    @Throws(ParseException::class)
    private fun processBlockExpiry(response: MwQueryResponse): Boolean {
        val blockExpiry = response.query()?.userInfo()?.blockexpiry()
        return when {
            blockExpiry.isNullOrEmpty() -> false
            "infinite" == blockExpiry -> true
            else -> {
                val endDate = DateUtil.iso8601DateParse(blockExpiry)
                val current = Date()
                endDate.after(current)
            }
        }
    }
}
